### ISSUE TITLE
AMQP.Net Lite to rabbitMQ with diacritics in queue name

### ISSUE THREAD

**Issue by xbl1** [Azure/amqpnetlite#300]
Hi!

Using Amqp.Net Lite from the nuget package or recompiled from the repository.

Broker is RabbitMQ 3.7.6 with Erlang 20.2.2, running on Ubuntu or on Windows.

In the HelloAmqp example from the sources, change the queue name to add diacritics, for instance "été".

The example works fine, the message is sent and received without errors.

But, if you go to the admin page of RabbitMQ, you'll notice that the diacritics have been replaced by unexpected characters. Furthermore, the command: 

rabbitmqctl list_queues

Throws an "invalid encoding" error.

Performing the same test with Pyhton+pika to the same RabbitMQ creates the queue with the expected diacritics, but a receiver using Amqp.Net Lite will not attach to this correct queue (and hence not receive the message).

Performing the same test with Amqp.Net Lite to ActiveMQ creates the queue with the expected diacritics. An Amqp.net lite receiver can be attached to the queue and will receive message sent from Pyhton.

So the problem lies in the interaction between Amqp.Net Lite and RabbitMQ, but I can't put my finger on it.

Do you have any clue about this problem?

Thanks in advance

**Comment by xbl1** [comment]
Hi!

Thanks for your answer. I performed the following additional tests.

1. Using AMQP.Net Lite, send a message to RabbitMQ with diacritics in the body and some properties (MessageId, replyTo...), but no diacritics in the queue name.
2. Still using AMQP.Net Lite, receive this message => body and properties are OK
3. Using python+pika, receive this same message => body and properties are OK

So to answer your question, diacritics in everything but the queue name works OK with RabbitMQ, including from AMQP.Net Lite to python+pika

Furthermore, as stated in my original message, using python+pika to send and receive a message to a queue name with diacritics works OK too. The problem only happens with AMQP.Net Lite and RabbitMQ.

As anyone ever met this situation?

Thanks in advance.

**Comment by xinchen10** [comment]
Yes I was referring to the Dynamic flag in Target/Source because your first post indicates that the queue was create somehow by the client request. Did you have to pre-create the queue? If not, was the queue auto-created even without the Dynamic flag? Sorry I am not familiar with RabbitMQ.

**Comment by xbl1** [comment]
No problem, I'm actually not very familiar with RabbitMQ either. Thanks for you time in dealing with this issue.

No, I didn't have to create the queue. It is automatically created (with the funky characters instead of the diacritics) by RabbitMQ.

Furthermore, if I anyway create the queue in RabbitMQ Web interface, with the diacritics, it is correctly created and displayed, but sending to it with AMQP.Net Lite will create another queue with the wrong characters.

I'm wondering if I can try to find the encoded strings in the transmitted frames (ideally without sniffing the newtork packets), to see if I can spot a difference between AMQP.Net Lite and Pika. I am currently investigating this option.

**Comment by xbl1** [comment]
Using tcpdump, I inspected the packets transmitted, looking for the bytes pattern, and I noticed something odd. These are the queue names in the packets in the two clients:

AMQP.NET Lite:
a1 05 c3 a9 74 c3 a9
Python Pika:
a0 00 00 05 c3 a9 74 c3 a9

AMQP.Net Lite is transmitting a correctly formatted str8-utf8 string (0xa1 as first byte), when Pika transmits a vbin8 byte buffer (0xa0 as first byte), but I cannot match it with the AMQP 1.0 specification. In my understanding, we should have a0 immediately followed by a one-byte length (it should look like a0 05 c3 a9 74 c3 a9). I may be mistaken, but all in all this is not the place to discuss the odds and ends of Pika, if any.

Nevertheless, it appears that AMQP.Net Lite sends a correctly formatted queue name, that is not handled properly by RabbitMQ, whereas Pika sends the queue name as a byte array (correctly constructed or not), that RabbitMQ accepts.

In any case, I  am now convinced that the problem comes from RabbitMQ, not from AMQP.Net Lite.

I think we can close this issue, except if you feel like doing something about it to enhance interoperability with RabbitMQ.

Thanks again for your time!
